test/functional/test-amp-context.js

```diff
@@ -0,0 +1,316 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {
+  AmpContext,
+} from '../../3p/ampcontext';
+import {MessageType} from '../../src/3p-frame-messaging';
+import * as sinon from 'sinon';
+
+describe('3p ampcontext.js', () => {
+  let windowPostMessageSpy;
+  let windowMessageHandler;
+  let win;
+  let sandbox;
+
+  beforeEach(() => {
+    sandbox = sinon.sandbox.create();
+    windowPostMessageSpy = sinon.sandbox.spy();
+    win = {
+      addEventListener: (eventType, handlerFn) => {
+        expect(eventType).to.equal('message');
+        expect(windowMessageHandler).to.not.be.ok;
```

Comment:
useful as a sanity check that we only try to test one message per test. If you try to register a second event listener in one test, as it is right now, it will error instead of just letting you overwrite the initial message handler.